### PR TITLE
Add a contextmanager to change the assumption that bool vectors are masks

### PR DESCRIPTION
**Pull request by isuruf**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #116491
* __->__ #126406

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @ColinPeppler @amjames @desertfire @chauhang

### CODE REVIEW

File: torch/_inductor/codegen/cpp.py [hunk 1]
```diff
@@ -1397,7 +1397,7 @@ def masked(mask, body, other):
         assert isinstance(V.kernel, CppVecKernel)
         code = BracesBuffer()
         var = V.kernel.cse.newvar()
-        with V.kernel.masked(mask) as new_mask:
+        with V.kernel.masked(mask) as new_mask, V.kernel.bool_vectors_are_masks(False):
```
Comment (peterbell10) [inline torch/_inductor/codegen/cpp.py:1400]: I don't think this will work if we have kernels that mix bools and floats or other vectorized dtypes. Can you give a minimal reproducer for the failure with the current codegen?

Comment (isuruf) [reply torch/_inductor/codegen/cpp.py:1400]:
```python
import torch
cfn = torch.compile(torch.constant_pad_nd)
cfn(torch.zeros(5, dtype=torch.bool), [2, 3])
```

Comment (isuruf) [reply torch/_inductor/codegen/cpp.py:1400]: You are right. This is not the correct solution. I'll open an issue
